### PR TITLE
RWA-1721: Granular Permission: Verify role assignments with granular permission requirements

### PR DESCRIPTION
**Pull request by rasikasri**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-1721

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Review by chrisdavecm — APPROVED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/ia/CftQueryServiceGetTaskWithGranularPermission.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import static uk.gov.hmcts.reform.wataskmanagementapi.auth.permission.entities.PermissionTypes.OWN;
+import static uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.search.parameter.SearchParameterKey.LOCATION;
+
+@ActiveProfiles("integration")
+@DataJpaTest
+@Import(AllowedJurisdictionConfiguration.class)
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@Testcontainers
+@Sql("/scripts/ia/get_task_granular_permission_data.sql")
+public class CftQueryServiceGetTaskWithGranularPermission extends RoleAssignmentHelper {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/ia/CftQueryServiceGetTaskWithGranularPermission.java:57]: Keep Test Class name consistent with naming convention. ClassNameTest for example

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskWithGranularPermission.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+    @BeforeEach
+    void setUp() {
+        CFTTaskMapper cftTaskMapper = new CFTTaskMapper(new ObjectMapper());
+        cftQueryService = new CftQueryService(camundaService, cftTaskMapper, new TaskResourceDao(entityManager),
+                                              allowedJurisdictionConfiguration
+        );
+    }
+
+    @Test
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskWithGranularPermission.java:70]: could we extend this so WA jurisdiction also caters for Specific and Standard? We are gradually moving away from IA being the only service that utilises WA/TM.

Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskWithGranularPermission.java:70]: This test class is for WA jurisdiction

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/ia/CftQueryServiceGetTaskWithGranularPermission.java (hunk 1 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/ia/CftQueryServiceGetTaskWithGranularPermission.java]: Done

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskWithGranularPermission.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+import static uk.gov.hmcts.reform.wataskmanagementapi.auth.permission.entities.PermissionTypes.EXECUTE;
+import static uk.gov.hmcts.reform.wataskmanagementapi.auth.permission.entities.PermissionTypes.OWN;
+
+@ActiveProfiles("integration")
+@DataJpaTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@Import(AllowedJurisdictionConfiguration.class)
+@Testcontainers
+@Sql("/scripts/wa/get_task_granular_permission_data.sql")
+public class CftQueryServiceGetTaskWithGranularPermission extends RoleAssignmentHelper {
```
[the remark this answers is not in the document]
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskWithGranularPermission.java]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java [hunk 4]
```diff
@@ -98,37 +173,45 @@ public static Predicate buildRoleAssignmentConstraints(
...
+            if (permissionRequirementPredicate == null) {
+                permissionRequirementPredicate = permissionPredicate;
+            } else if (PermissionJoin.AND.equals(nextPermissionJoin)) {
+                permissionRequirementPredicate = builder.and(permissionRequirementPredicate, permissionPredicate);
+            } else {
+                permissionRequirementPredicate = builder.or(permissionRequirementPredicate, permissionPredicate);
+            }
 
+            nextPermissionJoin = nextRequirements.getPermissionJoin();
+            nextRequirements = nextRequirements.getNextPermissionRequirements();
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:212]: Think I get what you're doing here with current/next etc.

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/ia/CftQueryServiceGetTaskWithGranularPermission.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/ia/CftQueryServiceGetTaskWithGranularPermission.java]: I already changed this.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java (hunk 4 above)
Comment (rasikasri) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:212]: Yes, here it loop through the collection of permission requirement.

**Review by chrisdavecm — APPROVED** [review]

**Pull request merged**